extensions/ql-vscode/src/databases.ts

```diff
@@ -341,28 +341,30 @@ export class DatabaseItemImpl implements DatabaseItem {
     }
   }
 
-  public resolveSourceFile(uri: string | undefined): vscode.Uri {
+  public resolveSourceFile(uriStr: string | undefined): vscode.Uri {
     const sourceArchive = this.sourceArchive;
-    // Sometimes, we are passed a path, sometimes a file URI.
-    // We need to convert this to a file path that is probably inside of a zip file.
-    const file = uri?.replace(/file:/, '');
+    const uri = uriStr ? vscode.Uri.parse(uriStr, true) : undefined;
+    if (uri && uri.scheme !== 'file') {
+      throw new Error(`Invalid uri scheme in ${uriStr}. Only 'file' is allowed.`);
+    }
     if (!sourceArchive) {
-      if (file) {
-        // Treat it as an absolute path.
-        return vscode.Uri.file(file);
+      if (uri) {
+        return uri;
       } else {
         return this.databaseUri;
       }
     }
 
-    if (file) {
-      const absoluteFilePath = file.replace(':', '_');
-      // Strip any leading slashes from the file path, and replace `:` with `_`.
-      const relativeFilePath = absoluteFilePath.replace(/^\/*/, '').replace(':', '_');
+    if (uri) {
+      // path or fsPath
+      const relativeFilePath = decodeURI(uri.path.replace(':', '_').replace(/^\/*/, ''));
```

Comment:
Replacement should happen after decoding here.

Reply:
Yes